client/windows/CCastleInterface.cpp

```diff
@@ -361,11 +437,16 @@ void CHeroGSlot::gesture(bool on, const Point & initialPosition, const Point & f
 		{ RadialMenuConfig::ITEM_EE, twoHeroes, "tradeHeroes", "vcmi.radialWheel.heroExchange", [heroId, heroOtherId](){LOCPLINT->showHeroExchange(heroId, heroOtherId);} },
 		{ RadialMenuConfig::ITEM_SW, twoHeroes, "moveArtifacts", "vcmi.radialWheel.heroGetArtifacts", [heroId, heroOtherId](){CExchangeController(heroId, heroOtherId).moveArtifacts(false, true, true);} },
 		{ RadialMenuConfig::ITEM_SE, twoHeroes, "swapArtifacts", "vcmi.radialWheel.heroSwapArtifacts", [heroId, heroOtherId](){CExchangeController(heroId, heroOtherId).swapArtifacts(true, true);} },
-		{ RadialMenuConfig::ITEM_WW, true, "dismissHero", "vcmi.radialWheel.heroDismiss", [this]()
+		{ RadialMenuConfig::ITEM_WW, true, !upgradableSlots.isCreatureUpgradePossible ? "dismissHero" : "upgradeCreatures", !upgradableSlots.isCreatureUpgradePossible ? "vcmi.radialWheel.heroDismiss" : "vcmi.radialWheel.upgradeCreatures", [this, upgradableSlots, upgradeAll]()
 		{
-			CFunctionList<void()> ony = [=](){ };
-			ony += [=](){ LOCPLINT->cb->dismissHero(hero); };
-			LOCPLINT->showYesNoDialog(CGI->generaltexth->allTexts[22], ony, nullptr);
+			if(upgradableSlots.isCreatureUpgradePossible)
+				upgradeAll();
+			else
+			{
+				CFunctionList<void()> ony = [=](){ };
+				ony += [=](){ LOCPLINT->cb->dismissHero(hero); };
+				LOCPLINT->showYesNoDialog(CGI->generaltexth->allTexts[22], ony, nullptr);
+			}
```

Comment:
Perhaps simplify code? This version is rather hard to read, does 2 different things, and has 3 identical copy-pasted checks.

Perhaps something like:
```cpp
std::vector<RadialMenuConfig> menuElements = {
 (five permanent elements)
};

if (upgradableSlots.isCreatureUpgradePossible)
    menuElements.push_back(upgradeSlot)
else
    menuElements.push_back(dismissSlot)